src/category_theory/limits/shapes/constructions/over/default.lean

```diff
@@ -19,19 +19,32 @@ variable {X : C}
 namespace category_theory.over
 
 /-- Make sure we can derive pullbacks in `over B`. -/
-example {B : C} [has_pullbacks C] : has_pullbacks (over B) :=
-{ has_limits_of_shape := infer_instance }
+example {B : C} [has_pullbacks C] : has_pullbacks (over B) := infer_instance
 
 /-- Make sure we can derive equalizers in `over B`. -/
-example {B : C} [has_equalizers C] : has_equalizers (over B) :=
-{ has_limits_of_shape := infer_instance }
+example {B : C} [has_equalizers C] : has_equalizers (over B) := infer_instance
 
 instance has_finite_limits {B : C} [has_finite_wide_pullbacks C] : has_finite_limits (over B) :=
 begin
   apply @finite_limits_from_equalizers_and_finite_products _ _ _ _,
-  { exact construct_products.over_finite_products_of_finite_wide_pullbacks },
+  { /-
+    FIXME why do we need to insert `_inst_3` by hand here? It should suffice to write just:
+    ```
+    exact construct_products.over_finite_products_of_finite_wide_pullbacks
+    ```
+    but that gives the unhelpful error message (with pp.all):
+    ```
+    failed to synthesize type class instance for
+    C : Type u,
+    _inst_2 : category_theory.category.{v u} C,
+    B : C,
+    _inst_3 : @category_theory.limits.has_finite_wide_pullbacks.{v u} C _inst_2
+    ⊢ @category_theory.limits.has_finite_wide_pullbacks.{v u} C _inst_2
+    ```
+    -/
+    exact @construct_products.over_finite_products_of_finite_wide_pullbacks _ _ _inst_3 _ },
```

Comment:
@gebner, could I get you to take a quick look at this? I'm worried it's a bug in typeclass inference. It seems strange to me that I'd get this "failed to synthesize" error message, when the thing it's looking for is right there as a local hypothesis.

Reply:
(`resetI` and friends don't help)

Reply:
I suspect the problem may be that `has_finite_wide_pullbacks` is defined as a pi-type, as
```
abbreviation has_finite_wide_pullbacks : Type (max (v+1) u) :=
Π (J : Type v) [decidable_eq J] [fintype J], has_limits_of_shape (wide_pullback_shape J) C
```
and somehow that is snarling things up.

However just below, a very similar inference for `has_wide_pullbacks` succeeds just fine, which is also a pi-type:
```
abbreviation has_wide_pullbacks : Type (max (v+1) u) :=
Π (J : Type v), has_limits_of_shape (wide_pullback_shape J) C
```

Perhaps something about the extra typeclass arguments in the pi-type for `has_finite_wide_pullbacks` is causing the problem?

Reply:
I'm suspecting this is a bug in Lean, perhaps somehow related to leanprover-community/lean#411. Here is an MWE that reproduces the bug:

```lean
class foo (n : ℕ) : Prop :=
(has_true : true)

class bar (n : ℕ) : Prop :=
(has_true : true)

abbreviation quux := ∀ n [foo n], bar n

lemma have_quux : quux := λ n hn, ⟨trivial⟩

example [quux] : quux :=
begin
  apply_instance,
/-
tactic.mk_instance failed to generate instance for
  quux
state:
_inst_1 : quux
⊢ quux
-/
end
```

Here are some things that make the problem go away:
* Remove the type class argument in the definition of `quux`
* Change `quux` into a `@[class] def`

Here are some things that do not make the problem go away:
* Change `quux` into a `@[class] abbreviation`
* Switch to Lean 3.17.1 (so maybe this isn't related to leanprover-community/lean#411)

Reply:
Even more surprisingly, the following fails back to 3.4.2:
```
example [∀ n [foo n], bar n] : quux := by apply_instance
example [∀ n [foo n], bar n] : ∀ n [foo n], bar n := by apply_instance
```

This works though:
```
example [∀ n [foo n], bar n] (n) [foo n] : bar n := by apply_instance
```

Reply:
I've filed this as https://github.com/leanprover-community/lean/issues/429

As mentioned in the issue, I don't think that this pattern is a good idea.  You're not supposed to introduce instances locally (without using `resetI` and friends).

Reply:
A potential workaround is the following:
```lean
def has_finite_wide_pullbacks : Type (max (v+1) u) :=
Π (J : Type v) [decidable_eq J] [fintype J], has_limits_of_shape (wide_pullback_shape J) C

instance [has_finite_pullbacks C] (J : Type v) [decidable_eq J] [fintype J] :
  has_limits_of_shape (wide_pullback_shape J) C :=
«has_finite_pullbacks C» J

-- other instances are fine as long as they don't have arguments like `[∀ [foo], bar]`
```

Reply:
I'll use @gebner's workaround, thanks!

Reply:
(Just in case anyone needs this, the workaround needs to be tweaked to:
```
def has_finite_wide_pullbacks : Type (max (v+1) u) :=
Π (J : Type v) [decidable_eq J] [fintype J], has_limits_of_shape (wide_pullback_shape J) C

instance (J : Type v) [decidable_eq J] [fintype J] [has_finite_wide_pullbacks C] :
  has_limits_of_shape (wide_pullback_shape J) C :=
‹has_finite_wide_pullbacks C› J
```

Reply:
All fixed up! Thanks very much @TwoFX for minimising, and @gebner for the workaround.